### ISSUE TITLE
Result of torch.autograd.grad(create_graph=True) sometimes has requires_grad=False

### ISSUE THREAD

**Issue by 212726320** [pytorch/pytorch#30346]
## 🐛 Bug

Typically, the `torch.Tensor` returned by `torch.autograd.grad(*args, create_graph=True)` will have `requires_grad=True`.  However, there seem to be some cases (see below) when the result has (unexpectedly) `requires_grad=False`.

I generally expect the former behavior all the time, but [the documentation for `torch.autograd.grad`](https://pytorch.org/docs/stable/autograd.html#torch.autograd.grad) does not actually seem to define what the intended behavior is.

As a side comment (not shown in MWE), adding some version of "zero times the input" to the output reconnects the gradient flow and seems to be a feasible workaround.

## To Reproduce

```
import torch
x = torch.zeros(1)
x.requires_grad_(True)
y = x ** 2

dy = torch.autograd.grad(y, x, create_graph=True)[0]
z = 2 * x

d2y = torch.autograd.grad(dy, x, create_graph=True)[0]
dz = torch.autograd.grad(z, x, create_graph=True)[0]

print(d2y.requires_grad)
print(dz.requires_grad)

d3y = torch.autograd.grad(d2y, x, create_graph=True)[0]
# d2z = torch.autograd.grad(dz, x, create_graph=True)[0]  # Throws an error

print(d3y.requires_grad)
```
Outputs:
```
True
False
False
```

## Expected behavior

Outputs:
```
True
True
True
```

## Environment

```
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Microsoft Windows 10 Pro
GCC version: (GCC) 7.3.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Quadro P1000
Nvidia driver version: 419.17
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.3
[pip] torch==1.3.1
[conda] blas                      1.0                         mkl    defaults
[conda] mkl                       2019.4                      245    defaults
[conda] mkl-service               2.3.0            py37hb782905_0    defaults
[conda] mkl_fft                   1.0.15           py37h14836fe_0    defaults
[conda] mkl_random                1.1.0            py37h675688f_0    defaults
[conda] pytorch                   1.3.1           py3.7_cuda101_cudnn7_0    pytorch
```

## Additional context

N/A

**Comment by SsnL** [comment]
This happens because dz/dx is always the constant 2, which is computed by `d out /dz * dz/dx`, multiplying two tensors that don't require gradient. So your variable named `dz` does not require gradient.

**Comment by albanD** [comment]
I agree with @SsnL that this Tensor should not require gradients as it is independent of all the leafs in your graph.
